text/3127-trim-paths.md

```diff
@@ -0,0 +1,351 @@
+- `all` and `true` - an alias for all of the above, also equivalent to supplying only `--remap-path-prefix` without `--remap-path-scope`.
+
+Debug information are written to split files when the separate codegen option `-C split-debuginfo=packed` or `unpacked` (whether by default or explicitly set).
+
+Note: this RFC is not a commitment to stabilizing all of these options; stabilization will evaluate each option and see if that option carries enough value to stabilize.
+
+## Cargo
+
+`trim-paths` is a profile setting which enables and controls the sanitisation of file paths in build outputs. It is a simplified version of rustc's `--remap-path-scope`. It takes a comma separated list of the following values:
+
+- `none` and `false` - disable path sanitisation
+- `macro` - sanitise paths in the expansion of `std::file!()` macro. This is where paths in embedded panic messages come from
+- `diagnostics` - sanitise paths in printed compiler diagnostics
+- `object` - sanitise paths in compiled executables or libraries
+- `all` and `true` - sanitise paths in all possible locations
+
+Note: this RFC is not a commitment to stabilizing all of these options; stabilization will evaluate each option and see if that option carries enough value to stabilize.
+
+It is defaulted to `none` for debug profiles, and `object` for release profiles. You can manually override it by specifying this option in `Cargo.toml`:
+```toml
+[profile.dev]
+trim-paths = "all"
+
+[profile.release]
+trim-paths = "none"
+```
+
+The default release profile setting (`object`) sanitises only the paths in emitted executable or library files. It always affects paths from macros such as panic messages, and in debug information
+  only if they will be embedded together with the binary (the default on platforms with ELF binaries, such as Linux and windows-gnu),
+  but will not touch them if they are in separate files (the default on Windows MSVC and macOS). But the path to these separate files are sanitised.
+
+If `trim-paths` is not `none` or `false`, then the following paths are sanitised if they appear in a selected scope:
+
+1. Path to the source files of the standard and core library (sysroot) will begin with `/rustc/[rustc commit hash]`.
+   E.g. `/home/username/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/result.rs` -> 
+   `/rustc/fe72845f7bb6a77b9e671e6a4f32fe714962cec4/library/core/src/result.rs`
+2. Path to the working directory will be stripped. E.g. `/home/username/crate/src/lib.rs` -> `src/lib.rs`.
+3. Path to packages outside of the working directory will be replaced with `[package name]-[version]`. E.g. `/home/username/deps/foo/src/lib.rs` -> `foo-0.1.0/src/lib.rs`
+
+Paths requiring sanitisation can be retrieved by build scripts at their execution time from the environment variable `CARGO_TRIM_PATHS`, in comma-separated format.
```

Comment:
I'm having a little difficulty understanding `CARGO_TRIM_PATHS`. Which paths exactly will be passed into it? If it is more than the package root, why are those paths being included? If it is just the package root, that is already available via `CARGO_MANIFEST_DIR` (or the current working directory). I would think the `OUT_DIR` scratch directory is likely what is wanted, anyways?

Would it be useful to instead just pass in the value of the `trim-paths` profile setting? That way, the script can decide which kind of path trimming to use? I think in https://github.com/rust-lang/rfcs/pull/3127#issuecomment-869199458 you mentioned that gcc and clang have separate controls for the different kinds.

If there is some reason to pass in extra paths to the script, I think `\x1f` would be a better separator to use instead of comma (to be consistent with CARGO_ENCODED_RUSTFLAGS).

Reply:
@ehuss I'd expect it to cover more than just the package root; it'll also include things like paths to rust source code, or crates referenced by path dependency.

Reply:
What kind of situations would those be useful? Build scripts don't have direct access to those sources. The only thing I can think of is possibly running `cargo` recursively (which doesn't work very well), but I suspect something like https://github.com/rust-lang/cargo/issues/11244 might provide more fidelity.

Reply:
`CARGO_TRIM_PATHS` is useful when the build script invokes another compiler that with a similar path sanitisation feature: "Passing the information down the build scripts so cc-rs could pass --remap-path-prefix to clang or gcc" from the [Cargo meeting](https://hackmd.io/@rust-cargo-team/H1iucap35#2022-07-12). For instance, the user can manually remap paths to C header files after discovering them in the binary.

Most options in `trim-paths` are not relevant to build scripts. In fact, I expect that build scripts only need to care about path sanitisation when `trim-paths` contains `object`. Build scripts cannot affect `std::file!()` for the `macro` option, or rustc diagnostics for the `diagnostics` option.

Reply:
`CARGO_TRIM_PATHS` will contain:
- Path to the sysroot
- Path to the working directory
- Path to the package root
- Path manually specified to rustc via `--remap-path-prefix` flag (the main rationale of this flag)

Sysroot path isn't provided in other build script environment variables, but this is only useful when the external compiler accesses Rust sysroot, I agree that this is not a common scenario.

For build scripts, the working directory == package root right? The doc appears to say that the build script always has the package root as the cwd, even if it's not the cwd of the user `cargo` command. In this case, these two paths are indeed identical to `CARGO_MANIFEST_DIR`
> CARGO_MANIFEST_DIR — The directory containing the manifest for the package being built (the package containing the build script). Also note that this is the value of the current working directory of the build script when it starts.

`--remap-path-prefix` is normally passed using `RUSTFLAGS`, so it will be in `CARGO_ENCODED_RUSTFLAGS`. However, the build script would have to parse this manually.

It is true that the build script can recover most of the information in `CARGO_TRIM_PATHS` from other variables, but this is cumbersome to do in a build script.

There is the additional question of what the desired behaviour of a build script is, given a `trim-paths` profile setting. We could, as you mentioned, provide build scripts with the `trim-paths` option and let them decide the remapping flags supplied to the external compiler, although this may be overcomplicated and potentially very few build script authors care about this granularity. "When in doubt, trim it" may be the right thing to do, meaning that everything in `CARGO_TRIM_PATHS` should be trimmed everywhere by the external compiler, even if it is not strictly necessary for the given `trim-paths` profile setting.

Reply:
Yea, the working directory is always the package root.

I suppose including extra paths would not be harmful. I'd be a little surprised if anything would need to remap the sysroot, for example, since I don't think something like `cc` would ever access anything from there.
I would also suggest including OUT_DIR then, as I think that would be the primary directory that needs remapping aside from the package root.

As for passing in the trim-paths profile, it seems like that would be important information to me. The `cc` crate would likely need to know if it needs to pass in `-fdebug-prefix-map` or `-fmacro-prefix-map` with the same motivation as to why rustc has separate flags (and Cargo having the logic of deciding which flags to pass in). Without that information, it wouldn't be able to make that decision would it? If there is some reason that we don't want `cc` to make that distinction, then the reasoning should probably be written down. It seems surprising to me since I would expect `cc` to want to make the distinction for the same reason that `rustc` is.

> Path manually specified to rustc via --remap-path-prefix flag (the main rationale of this flag)

Cargo doesn't know what these are. It doesn't interpret RUSTFLAGS. This is suggesting adding something that we've explicitly tried to avoid. I'm not saying that this is not something that we can change, but it is a big change in policy that should probably be made clear. I would almost say that if it is intended for the user to pass in remappings manually, then there should be first-class support for that (not using RUSTFLAGS). If most of the remapping is being handled automatically by Cargo, what situations would be left for the user to pass things in manually?

Reply:
> It doesn't interpret RUSTFLAGS. This is suggesting adding something that we've explicitly tried to avoid.

Yeah I see your point. It's not desirable to make Cargo parse one particular flag in `RUSTFLAGS`, at least not in this RFC.

> what situations would be left for the user to pass things in manually

The first thing that comes to mind is include paths.

If we can't add manually specified `--remap-path-prefix` to `CARGO_TRIM_PATHS`, then I agree that the build script should determine the sanitised paths itself using other variables like `OUT_DIR` and `CARGO_MANIFEST_DIR`, plus any other paths it knows about.

In this case, I'll change `CARGO_TRIM_PATHS` from containing paths to containing the profile setting of `trim-paths`, and add docs telling build script authors to take care of paths introduced by external programs